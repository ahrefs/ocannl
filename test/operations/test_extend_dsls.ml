open! Base
open Ocannl.Operation.DSL_modules

(* Test module-level expansion *)
[%%extend_dsls
let my_tensor x = TDSL.O.( + ) x !.2.0 + !.1.0

let my_complex_op a b c =
  let d = a * b in
  let e = d + c in
  relu e

let my_random_op () = { u1 = uniform () } - { u2 = uniform () }]

(* The above should create three modules: TDSL, NTDSL, and PDSL, each containing the functions with
   appropriate DSL operators *)

let () =
  (* Test that the functions are available in each DSL module *)
  let open! DSL_modules.TDSL.O in
  let x = uniform () in
  let result = my_complex_op (my_tensor x) (my_random_op ()) x in
  Tensor.print ~here:[%here] ~force:false ~with_code:false ~with_grad:true `Inline result

let () =
  let open! DSL_modules.NTDSL.O in
  let x = uniform () in
  let result = my_complex_op (my_tensor x) (my_random_op ()) x in
  Tensor.print ~here:[%here] ~force:false ~with_code:false ~with_grad:true `Inline result

let () =
  let open! DSL_modules.PDSL.O in
  let x = uniform () in
  let result = my_complex_op (my_tensor x) (my_random_op ()) x in
  Tensor.print ~here:[%here] ~force:false ~with_code:false ~with_grad:true `Inline result
