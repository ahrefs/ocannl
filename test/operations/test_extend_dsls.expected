Retrieving commandline, environment, or config file variable ocannl_log_level
Found 0, in the config file
HERE: test/operations/test_extend_dsls.ml:23:21
[23]: relu shape 0:$80  <not-in-yet>

HERE: test/operations/test_extend_dsls.ml:29:21
[46]: relu shape 0:$161  <not-in-yet>

HERE: test/operations/test_extend_dsls.ml:35:21
[75]: relu shape 0:$242  <not-in-yet>
grad_relu <not-in-yet>
